packages/api/src/utils/stripe.js

```diff
@@ -396,13 +397,196 @@ export function createStripeBillingContext (env) {
+  /** @type {UserCustomerService} */
   const userCustomerService = {
     upsertUserCustomer: env.db.upsertUserCustomer.bind(env.db),
     getUserCustomer: env.db.getUserCustomer.bind(env.db)
   }
   const customers = StripeCustomersService.create(stripe, userCustomerService)
+  const subscriptions = StripeSubscriptionsService.create(stripe)
   return {
     billing,
-    customers
+    customers,
+    subscriptions
   }
 }
+
+/**
+ * @typedef {object} StripeApiForSubscriptionsService
+ * @property {Pick<Stripe['subscriptions'], 'cancel'|'create'>} subscriptions
+ * @property {Pick<Stripe['subscriptionItems'], 'update'|'del'>} subscriptionItems
+ * @property {Pick<Stripe['customers'], 'retrieve'>} customers
+ */
+
+/**
+ * A SubscriptionsService that uses stripe.com for storage
+ */
+export class StripeSubscriptionsService {
+  /**
+   * @param {StripeApiForSubscriptionsService} stripe
+   */
+  static create (stripe) {
+    return new StripeSubscriptionsService(stripe)
+  }
+
+  /**
+   * @param {StripeApiForSubscriptionsService} stripe
+   * @protected
+   */
+  constructor (stripe) {
+    /** @type {StripeApiForSubscriptionsService} */
+    this.stripe = stripe
+    /**
+     * @type {import('./billing-types').SubscriptionsService}
+     */
+    const instance = this // eslint-disable-line
+  }
+
+  /**
+   * @param {string} customerId
+   * @returns {Promise<import('./billing-types').W3PlatformSubscription|CustomerNotFound>}
+   */
+  async getSubscription (customerId) {
+    const storageStripeSubscription = await this.getStorageStripeSubscription(customerId)
+    if (storageStripeSubscription instanceof CustomerNotFound) { return storageStripeSubscription }
+    /** @returns {import('./billing-types').W3PlatformSubscription} */
+    const subscription = {
+      storage: createW3StorageSubscription(storageStripeSubscription)
+    }
+    return subscription
+  }
+
+  async getStorageStripeSubscription (customerId) {
+    const customer = await this.stripe.customers.retrieve(customerId, {
+      expand: ['subscriptions']
+    })
+    if (customer.deleted) {
+      return new CustomerNotFound('customer retrieved from stripe has been unexpectedly deleted')
+    }
+    const { subscriptions: stripeSubscriptions } = customer
+    if (!stripeSubscriptions) {
+      // this is unexpected, since we requested expand=subscriptions above
+      throw new Error('expected subscriptions to be expanded, but got falsy value')
+    }
+    const storageStripeSubscription = selectStorageStripeSubscription(customerId, stripeSubscriptions)
+    return storageStripeSubscription
+  }
+
+  /**
+   *
+   * @param {import('./billing-types').CustomerId} customerId
+   * @param {import('./billing-types').W3PlatformSubscription} subscription
+   * @returns {Promise<CustomerNotFound|void>}
+   */
+  async saveSubscription (customerId, subscription) {
+    const storageStripeSubscription = await this.getStorageStripeSubscription(customerId)
```

Comment:
This feels a bit gross to check whether there is already a subscription for the customer by going to stripe.com APIs.
1. its slower than checking our db
2. there's a race condition unless we use a db transaction (two concurrent requests will both see no previous subscription so both try to create one)
  * but even if I use the db... our postgrest/supbase setup doesn't support transactions I think. So it's mostly about #1 IMO

Alternatively we could keep track of the subscription in our db through a new [customer_id, subscription_id] 'join' table. But there are some other must-dos to get this ready for working across our separate staging/prod stripe accounts, so I want to do that first.

Is this approveable as is without making a db schema changes to keep track of subscription on our side?

Reply:
@gobengo and I just talked about this and agreed that it's a better use of time to make the stripe fixtures consistent between staging & prod envs, given that the race condition is likely to be rare and we can recover from an error with a manual fix if needed. Definitely worth adding the join table when things settle down though.